source/common/config/incremental_subscription_impl.h

```diff
@@ -0,0 +1,312 @@
+    if (stream_ == nullptr) {
+      ENVOY_LOG(warn, "Unable to establish new stream");
+      onIncrementalConfigFailed(nullptr);
+      setRetryTimer();
+      return;
+    }
+
+    // "must be populated for first request in a stream"
+    request_.clear_initial_resource_versions();
+    for (auto const& resource : resources_) {
+      (*request_.mutable_initial_resource_versions())[resource.first] = resource.second;
+    }
+
+    control_plane_stats_.connected_state_.set(1);
+    queueDiscoveryRequest();
+  }
+
+  void drainRequests() {
+    ENVOY_LOG(trace, "draining discovery requests {}", request_queue_.size());
+    while (!request_queue_.empty()) {
+      // Process the request, if rate limiting is not enabled at all or if it is under rate limit.
+      if (!rate_limiting_enabled_ || limit_request_->consume()) {
+        sendDiscoveryRequest();
+      } else {
+        ASSERT(rate_limiting_enabled_);
+        ASSERT(drain_request_timer_ != nullptr);
+        control_plane_stats_.rate_limit_enforced_.inc();
+        control_plane_stats_.pending_requests_.set(request_queue_.size());
+        drain_request_timer_->enableTimer(
+            std::chrono::milliseconds(limit_request_->nextTokenAvailableMs()));
+        break;
+      }
+    }
+  }
+
+  // Enqueues and attempts to send a discovery request with no change to subscribed resources.
+  void queueDiscoveryRequest() {
+    request_queue_.emplace();
+    drainRequests();
+  }
+
+  // Enqueues and attempts to send a discovery request, (un)subscribing to resources missing from /
+  // added to the passed 'resources' argument, relative to resources_. Updates resources_ to
+  // 'resources'.
+  void queueDiscoveryRequest(const std::vector<std::string>& resources) {
+    ResourceNameDiff diff;
+    for (const auto& resource : resources) {
+      if (resources_.find(resource) == resources_.end()) {
+        diff.added_.push_back(resource);
+      }
+    }
+    for (const auto& entry : resources_) {
+      bool found = false;
+      for (const auto& resource : resources) {
+        if (entry.first == resource) {
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        diff.removed_.push_back(entry.first);
+      }
+    }
+    for (const auto& added : diff.added_) {
+      resources_[added] = "0";
+    }
+    for (const auto& removed : diff.removed_) {
+      resources_.erase(removed);
+    }
+    drainRequests();
+  }
+
+  void sendDiscoveryRequest() {
+    if (stream_ == nullptr) {
+      // Don't immediately try to reconnect: we rely on retry_timer_ for that.
+      ENVOY_LOG(debug, "No stream available to sendDiscoveryRequest for {}", type_url_);
+      return;
+    }
+    if (paused_) {
+      ENVOY_LOG(trace, "API {} paused during sendDiscoveryRequest().", type_url_);
+      return;
+    }
+
+    request_.clear_resource_names_subscribe();
+    request_.clear_resource_names_unsubscribe();
+    const ResourceNameDiff& diff = request_queue_.front();
```

Comment:
Yeah, the more I look at this, the more I'm seeing a lot of complicated code around rate limiting that should be collapsed into a shared implementation. It would also make the incremental code more readable, since it would be easier to focus on just the diffing aspects.